src/components/legend/draw.js

```diff
@@ -335,18 +336,15 @@ module.exports = function draw(gd) {
                 } else {
                     var traces = [],
                         clickedTrace;
-                    fullLayout._infolayer.selectAll('g.traces').each(function() { d3.select(this).call(function() { traces.push(this); }); });
-                    for(var i = 0; i < traces.length; i++) {
-                        var tracei = traces[i];
-                        var p = tracei[0][0].getBoundingClientRect();
-                        if(e.clientX >= p.left && e.clientX <= p.right && e.clientY >= p.top && e.clientY <= p.bottom) {
-                            clickedTrace = tracei;
-                            break;
-                        }
-                    }
-                    if(clickedTrace) {
+                    traces = fullLayout._infolayer.selectAll('g.traces').filter(function() {
+                        var bbox = this.getBoundingClientRect();
+                        return (e.clientX >= bbox.left && e.clientX <= bbox.right &&
+                            e.clientY >= bbox.top && e.clientY <= bbox.bottom);
+                    })[0];
+                    if(traces.length > 0) {
+                        clickedTrace = d3.select(traces[0]);
```

Comment:
non-blocking, but I find it confusing to dig into d3's nested arrays directly. Preferable to leave `traces` as a selection, then use `traces.size()` instead of `traces[0].length`, and then you don't need to re-select it to make `clickedTrace`.